<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="Jon Badalamenti">
        
        <link rel="shortcut icon" href="./img/favicon.ico">

	<title>Tn-seq</title>

        <link href="./css/bootstrap-3.0.3.min.css" rel="stylesheet">
        <link href="./css/font-awesome-4.0.3.css" rel="stylesheet">
        <link rel="stylesheet" href="./css/highlight.css">
        <link href="./css/base.css" rel="stylesheet">
        <link href="./extra.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href=".">Tn-seq</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
                <li class="active">
                    <a href=".">Home</a>
                </li>
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                <li class="disabled">
                    <a rel="next" >
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li class="disabled">
                    <a rel="prev" >
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
                <li>
                    <a href="https://github.com/jbadomics/tnseq">
                        
                            <i class="fa fa-github"></i>
                        
                        GitHub
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#identifying-harmful-mutations-in-microbial-populations-with-tn-seq">Identifying harmful mutations in microbial populations with Tn-seq</a></li>
        
            <li><a href="#set-up-amazon-instance-and-install-dependencies">Set up Amazon instance and install dependencies</a></li>
        
            <li><a href="#introduction-to-tn-seq">Introduction to Tn-seq</a></li>
        
            <li><a href="#getting-down-to-base-ics">Getting down to base-ics</a></li>
        
            <li><a href="#tn-seq-data-analysis-workflow">Tn-seq data analysis workflow</a></li>
        
    
    </ul>
</div></div>
            <div class="col-md-9" role="main">

<h1 id="identifying-harmful-mutations-in-microbial-populations-with-tn-seq">Identifying harmful mutations in microbial populations with Tn-seq</h1>
<p>This repository contains lesson materials, instructions, and scripts for analyzing Tn-seq data as presented during the <a href="http://dib-training.readthedocs.org/en/pub/2016-02-08-bodega.html">Bodega Bay 2016 bioinformatics course</a>.</p>
<p><strong>DISCLAIMER:</strong> This lesson is written for the particular flavor of Tn-seq data generation and analysis that I am familiar with in my lab as described <a href="http://onlinelibrary.wiley.com/doi/10.1111/j.1365-2958.2012.08196.x/abstract">here</a>, but modifications to the general Tn-seq scheme exist. Also, keep in mind that, as was determined at #ngs2015</p>
<p><img alt="I am not a bioinformatician" src="./not_a_bioinformatician.jpg" /></p>
<h2 id="set-up-amazon-instance-and-install-dependencies">Set up Amazon instance and install dependencies</h2>
<p>Before we get going with data analysis, we need to set up our environment and install some dependencies. On Amazon Web Services, launch Ubuntu 14.04 LTS (64-bit) on an m3.2xlarge instance. <strong>Before you click 'Review and Launch'</strong>, increase the root file system size to 30 GB. You will need to create a new private key if you do not already have one.</p>
<p>When the instance becomes available, copy its address, open a Terminal and connect to it via ssh:</p>
<pre><code>ssh -i keyfile.pem ubuntu@copy.address.here.amazonaws.com
</code></pre>
<p>Now let's install software:</p>
<pre><code>sudo apt-get update &amp;&amp; \
sudo apt-get -y upgrade &amp;&amp; \
sudo apt-get -y install autoconf automake bison build-essential default-jdk default-jre expat fastqc fastx-toolkit  g++ gcc git libboost-all-dev libbz2-dev libncurses5-dev libpcre++-dev libpcre3-dev make parallel python-dev python-setuptools trimmomatic unzip wget zlib1g-dev
</code></pre>
<p>While I introduce my research and give an overview of Tn-seq, download and unzip the data file we'll be using:</p>
<pre><code>mkdir ~/data
sudo mkfs.ext4 -E nodiscard /dev/xvdc
sudo mount /dev/xvdc ~/data
sudo chown ubuntu ~/data
cd ~/data
wget http://dib-training.ucdavis.edu.s3.amazonaws.com/2016-bodega/tnseq_reads.fastq.gz
gunzip tnseq_reads.fastq.gz &amp;&amp; md5sum tnseq_reads.fastq.gz
</code></pre>
<p>This lesson is a bit of a disk hog, so let's mount the other SSD included with the m3.2xlarge instance to an analysis directory:</p>
<pre><code>sudo umount /dev/xvdb
mkdir ~/analysis
sudo mount /dev/xvdb ~/analysis
sudo chown ubuntu ~/analysis
</code></pre>
<p>We will be using some other software packages that require manual installation. First, we'll install Heng Li's <a href="https://github.com/lh3/bioawk">bioawk</a>, an extension of the powerful GNU <code>awk</code> language which readily parses and manipulates common bioinformatics file formats like fastx and sam:</p>
<pre><code>sudo mkdir /sw 
sudo chown ubuntu /sw
chmod 775 /sw
cd /sw
git clone https://github.com/lh3/bioawk.git
cd bioawk
make
</code></pre>
<p>Next, install <a href="https://github.com/bcthomas/pullseq">pullseq</a>. I've found it to be a really handy tool for grabbing reads from fastx files by name or by matching a regular expression:</p>
<pre><code>cd /sw
git clone https://github.com/bcthomas/pullseq.git
cd pullseq
./bootstrap
./configure
make
sudo make install
</code></pre>
<p>Install Trimmomatic:</p>
<pre><code>cd /sw &amp;&amp; wget http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.35.zip
unzip Trimmomatic-0.35.zip
echo 'trimmomatic=/sw/Trimmomatic-0.35/trimmomatic-0.35.jar' &gt;&gt; ~/.bashrc
</code></pre>
<p>Now let's install <a href="https://github.com/samtools/samtools/releases/tag/1.2">samtools</a> <a href="https://twitter.com/pathogenomenick/status/696409415302430721">version 1.2</a>:</p>
<pre><code>cd /sw
wget https://github.com/samtools/samtools/releases/download/1.2/samtools-1.2.tar.bz2
tar -xvjf samtools-1.2.tar.bz2
cd samtools-1.2
make
</code></pre>
<p>Install <a href="https://sourceforge.net/projects/bowtie-bio/files/bowtie/1.1.2/">bowtie</a>:</p>
<pre><code>cd /sw
wget https://sourceforge.net/projects/bowtie-bio/files/bowtie/1.1.2/bowtie-1.1.2-src.zip
cd bowtie-1.1.2
make
</code></pre>
<p>Make sure <code>bash</code> knows where we've installed our packages:</p>
<pre><code>echo 'PATH=~/tnseq/scripts:$PATH' &gt;&gt; ~/.bashrc
echo 'PATH=/sw:$PATH' &gt;&gt; ~/.bashrc
echo 'PATH=/sw/samtools-1.2:$PATH' &gt;&gt; ~/.bashrc
echo 'PATH=/sw/bowtie-1.1.2:$PATH' &gt;&gt; ~/.bashrc
source ~/.bashrc
which bioawk
</code></pre>
<p>We also need to install <a href="http://biopython.org/wiki/Main_Page">BioPython</a>:</p>
<pre><code>sudo easy_install pip setuptools
sudo pip install --upgrade pip setuptools
sudo -H pip install pyopenssl ndg-httpsclient pyasn1
sudo -H pip install biopython
</code></pre>
<p>Finally, clone the lesson repo into your home directory:</p>
<pre><code>cd ~
git clone https://github.com/jbadomics/tnseq.git
</code></pre>
<h2 id="introduction-to-tn-seq">Introduction to Tn-seq</h2>
<p>In this lesson, we'll be analyzing Tn-seq data from an environmental bacterium that we study in the <a href="http://thebondlab.org">Bond Lab</a>, <em>Geobacter sulfurreducens</em>. This organism is a model system for microbial metal reduction and extracellular electron transfer, since it obtains metabolic energy by respiring insoluble metal oxides like Fe and Mn located outside the cell. Our lab is interested in identifying the proteins involved in this remarkable ability to transfer electrons across two insulating biological membranes.</p>
<p>Tn-seq is a hypothesis-generating tool for identifying genes that provide fitness benefit under particular conditions. The procedure requires the following:</p>
<ol>
<li>A genetically tractable model organism, and</li>
<li>its genome sequence.</li>
</ol>
<p>First, cells are randomly mutagenized with a suicide plasmid containing the Mariner transposon (which confers resistance to kanamycin). This transposon is flanked by inverted repeats and integrates into the chromosome at TA sites. A saturated mutant library is created by pooling a number of colonies roughly 10 times the number of genes in the genome. This number depends on two factors:</p>
<ul>
<li>genome size</li>
<li>GC content (higher GC microbes inherently have fewer TA sites)</li>
</ul>
<p>The transposon mutant library we'll be looking at contains ~40,000 individual mutants. Some of them will have no phenotype; others will be lethal (i.e. the transposon insertion has interrupted an essential gene); or, the insertion will cause a fitness defect only under certain conditions.</p>
<p>The entire mutant library is then subjected to two or more different outgrowth conditions, ideally for a known number of cell doublings (more on this later!). Assume a mutation at locus X, which causes a fitness defect under condition 1 but not condition 2. Over a few doubling events under condition 1, cells which carry the locus X mutation will be outcompeted and eventually die off; over the same period under condition 2, cells can continue to grow despite the locus X mutation. After outgrowth, genomic DNA is harvested and submitted for Illumina sequencing to locate the genomic position where the transposon inserted. In condition 1, very few (if any reads) will map, since the mutation caused a fitness defect, whereas in condition 2, more reads will map. Genes with the highest ratios of reads mapped can then be deleted and the phenotype can be verified.</p>
<h2 id="getting-down-to-base-ics">Getting down to base-ics</h2>
<p>Let's look at the molecular steps again. <img alt="tnseq workflow" src="./tnseq.png" /></p>
<p>The original TA site in the genome is duplicated during the transposition event. In addition, a single transposition event will result (theoretically) in two reads that should map to the same insertion site - one on the forward strand, and one on the reverse strand. We will count this as two hits at one site.</p>
<p>But there's a problem: when it comes time to map reads, any read mapping to the reverse strand will need to have its insertion position corrected. Consider the following example of two mock reads mapped to the first TA site in the genome. Here is the resulting SAM file:</p>
<pre><code>@HD VN:1.0  SO:unsorted
@SQ SN:chr1 LN:3820884
@PG ID:bowtie2  PN:bowtie2  VN:2.2.6    CL:&quot;bowtie2-align-s --wrapper basic-0 -x MN1 -S test.sam -q -U simulated.fastq&quot;
fwd_read    0   chr1    29  42  20M *   0   0   TATGGAAGAAGTTTGGCTCC    DFEEEEIDDDDDDDDDDDDD
rev_read    16  chr1    11  42  20M *   0   0   TCCCGAGAAGGTCTGGTTTA    DDDDDDDDDDDDDIEEEEFD
</code></pre>

<p>In column 4, we can see that the reverse read reports an alignment beginning at base 11, when in fact the TA site of insertion is at the end of the read. For this lesson, I define the coordinate of the TA site as the coordinate of the 'A' residue, which in this case is 30. We'll incorporate a step in our workflow to ensure that the forward and reverse reads arising from the same insertion event end up with an identical insertion coordinate reported.</p>
<h2 id="tn-seq-data-analysis-workflow">Tn-seq data analysis workflow</h2>
<h3 id="remove-phix-reads">Remove phiX reads</h3>
<p>In situations where Illumina reads are generated from the same DNA template (e.g. conserved regions of the 16S rRNA gene), it can be hard to differentiate clusters on the Illumina flow cell unless an external control is spiked into the sample, usually phage phiX DNA. Your sequencing provider may <em>say</em> they've removed phiX reads, but let's check just to be safe.</p>
<p>First, create a data analysis directory:</p>
<pre><code>mkdir ~/analysis &amp;&amp; cd ~/analysis
</code></pre>
<p>We'll use bowtie to map our Tn-seq reads:</p>
<pre><code>bowtie-build ~/tnseq/reference/phiX.fasta phiX
bowtie -q -S -p $(nproc --all) phiX ~/data/tnseq_reads.fastq phiX.sam
</code></pre>
<p>Have a look at the sam file:</p>
<pre><code>less -S phiX.sam
</code></pre>
<p>You'll notice that we have quite a few reads mapping to phiX. We should remove them from the dataset before we continue:</p>
<pre><code>samtools view -f 4 phiX.sam | cut -f1 | pullseq -i Tn-seq.fastq -N - &gt; phiX_removed.fastq
</code></pre>
<p>Let's break this down: <code>samtools view</code> with <code>-f 4</code> collects any <em>unmapped</em> reads, in sam format. The first column contains the read IDs. These are piped into <code>pullseq</code> using <code>-N</code> which takes the read names from STDIN (<code>-n</code> if read IDs are in another file).</p>
<p>...or not. Unfortunately during the lesson we quickly discovered that pullseq is a memory hog and will eventually run out of RAM. The simplest option is to re-run bowtie and ask it to return the unmapped reads:</p>
<pre><code>bowtie -q -S -p $(nproc --all) --un phiX_removed.fastq phiX ~/data/tnseq_reads.fastq phiX.sam
</code></pre>
<p>Alternatively, you can run this command after bowtie to pull the unmapped reads in FASTQ format:</p>
<pre><code>bioawk -c sam '{ if ($flag==4) {print "@"$qname; print $seq; print "+"; print $qual}}' phiX.sam &gt; phiX_removed.fastq
</code></pre>
<p>In this lesson I included a shell script called <code>countseq</code> which runs <code>bioawk</code> to correctly count the number of sequences in any fastq or fasta file. Make sure that <code>phiX_removed.fastq</code> contains fewer reads than our raw data:</p>
<pre><code>countseq *.fastq
</code></pre>
<h3 id="remove-transposon-sequence-filter-and-dereplicate">Remove transposon sequence, filter, and dereplicate</h3>
<p>Use <code>less</code> to have a look at the phiX-removed reads. You should see some patterns: are there multiple barcodes? Do you see any TA insertion sites? Do the 3' ends of the reads look similar?</p>
<p>Previous implementations of this workflow used <code>fastx_clipper</code>, part of the fastx toolkit, to remove transposon sequence at the 3' end of the read. Unfortunately this was painfully slow.</p>
<p>Trimmomatic can do what we want, and is WAY faster. Rather than trim off Illumina adaptors, we can specify a custom file with our transposon sequence to trim:</p>
<pre><code>java -Xmx28g -jar $trimmomatic SE -phred33 phiX_removed.fastq tn_removed.fastq ILLUMINACLIP:/home/ubuntu/tnseq/reference/tnseq_adapters.fa:3:30:10 MINLEN:16
</code></pre>
<p>Time to introduce the power of <code>bioawk</code>. I'm a huge fan of bioawk. We can use <code>awk</code>-like language to construct <code>if</code> statements, match regex patterns, and print reads meeting our criteria in either fasta or fastq format. In addition, bioawk automatically recognizes and parses these file formats (along with others like GFF and SAM) and assigns logical variables like <code>$seq</code> to describe the sequence and <code>$qual</code> to define the quality string.</p>
<p>Now we need to discard any reads that did not contain transposon sequence. By this point in our workflow, these reads can be distinguished as still being full-length (51 bp):</p>
<pre><code>bioawk -c fastx '{ if(length($seq) != 51) { print "@"$name; print $seq; print "+"; print $qual; }}' tn_removed.fastq &gt; tn_removed.filtered.fastq
</code></pre>
<p>Let's use bioawk again in a one-liner to capture a length distribution of the filtered reads:</p>
<pre><code>bioawk -c fastx '{ print length($seq) }' tn_removed.filtered.fastq | sort | uniq -c
</code></pre>
<p>As we expect, the overwhelming majority of our data is 20 or 21 bp. Keep in mind that the reads still contain barcode, which we'll remove shortly. But before we get to that, run bioawk to collect the reads of desired length:</p>
<pre><code>bioawk -c fastx '{ if(length($seq) &gt;= 19 &amp;&amp; length($seq) &lt; 23) { print "@"$name; print $seq; print "+"; print $qual; }}' tn_removed.filtered.fastq &gt; tn_removed.filtered.length.fastq
</code></pre>
<p>Rerun <code>countseq</code> to keep an eye on how many sequences we're discarding at each step:</p>
<pre><code>countseq tn_removed*.fastq
</code></pre>
<p>When we go to map reads against the reference genome, it is critical to keep track of the genomic position where the insertion occurred. Conceptually it makes sense to reverse complement the reads so that the TA insertion sequence occurs at the 5' end:</p>
<pre><code>fastx_reverse_complement -i tn_removed.filtered.length.fastq -o tn_removed.filtered.length.rc.fastq
</code></pre>
<p>What if a read doesn't begin with TA? This can happen sometimes, since the transposon does integrate at non-TA sites at ~2% frequency. We can safely remove the non-TA insertion reads with--you guessed it--bioawk:</p>
<pre><code>bioawk -c fastx '{ if ($seq ~ /^TA/) { print "@"$name; print $seq; print "+"; print $qual; }}' tn_removed.filtered.length.rc.fastq &gt; \ tn_removed.filtered.length.rc.TAonly.fastq
</code></pre>
<p>Just to be safe, let's write the reads that <em>don't</em> begin with TA to another file:</p>
<pre><code>bioawk -c fastx '{ if ($seq !~ /^TA/) { print "@"$name; print $seq; print "+"; print $qual; }}' tn_removed.filtered.length.rc.fastq &gt; nonTA.fastq
</code></pre>
<p>Take another peek at the reads with <code>less</code>: they should now all begin with TA and the barcode should now appear at the 3' end.</p>
<p>Time to separate reads by barcode. This dataset uses 3 barcodes:</p>
<pre><code>BC1 CAGT parent 
BC2 GACT high potential electrode outgrowth
BC3 GTGT low potential electrode outgrowth
</code></pre>

<p>Note that some reads end with N but should be considered as having the barcode. Bioawk can also take a regular expression:</p>
<pre><code>bioawk -c fastx '{ if ($seq ~ /CAG[TN]$/) { print "@"$name; print $seq; print "+"; print $qual; }}' tn_removed.filtered.length.rc.TAonly.fastq &gt; BC1.fastq
</code></pre>
<p>Change the regex pattern and run the same command for BC2:</p>
<pre><code>bioawk -c fastx '{ if ($seq ~ /GAC[TN]$/) { print "@"$name; print $seq; print "+"; print $qual; }}' tn_removed.filtered.length.rc.TAonly.fastq&gt; BC2.fastq
</code></pre>
<p>And for BC3:</p>
<pre><code>bioawk -c fastx '{ if ($seq ~ /GTG[TN]$/) { print "@"$name; print $seq; print "+"; print $qual; }}' tn_removed.filtered.length.rc.TAonly.fastq&gt; BC2.fastq
</code></pre>
<p>Make sure everything adds up:</p>
<pre><code>countseq tn_removed*.fastq
</code></pre>
<p>Now, use a <code>for</code> loop to remove the barcode from each dataset:</p>
<pre><code>for j in {1..3}; do fastx_trimmer -t 4 -i BC${j}.fastq -o BC${j}_barcode_removed.fastq; done
</code></pre>
<h3 id="map-reads-and-collect-hit-statistics">Map reads and collect hit statistics</h3>
<p>Thinking ahead: when it comes time to tabulate the insertion coordinates by locus tag, the reference genome name and sequence need to match what we use to map reads to. <strong>Unfortunately this isn't always the case.</strong> The safest bet is to pull the genome sequence from the same Genbank file we'll use later for tabulating insertions. Run this little python script:</p>
<pre><code>gbk_to_fasta.py Geobacter_sulfurreducens_MN1.gbk
</code></pre>
<p>Now the fun can begin! Use bowtie to map reads:</p>
<pre><code>bowtie-build Geobacter_sulfurreducens_MN1.fasta MN1
for j in {1..3}; do bowtie -q -p $(nproc --all) -S -n 0 -e 70 -l 28 --nomaqround -y -k 1 -a -m 1 --best --un BC${j}_unmapped.fastq MN1 BC${j}_barcode_removed.fastq BC${j}.sam; done
</code></pre>
<p>Take a peek at the output sam files:</p>
<pre><code>less -S BC1.sam
</code></pre>
<p>and look at a read that mapped to the reverse strand (look for bitwise flag 16). You'll notice that the read ends in TA, which it should if it mapped to the reverse strand, but the coordinate in column 4 refers to the <em>first</em> base aligned reading from left to right. This means we'll need to correct the coordinate for reads mapped to the minus strand so that they have the same coordinate (i.e. TA insertion site) as the forward read that arose from the same insertion event.</p>
<p>How's this for a one-liner:</p>
<pre><code>for j in {1..3}; do bioawk -c sam '{ if($flag==0) print $qname,$rname,$flag,$pos,$pos+1,length($seq); if($flag==16)  print $qname,$rname,$flag,$pos,$pos+length($seq)-1,length($seq) }' BC${j}.sam | cut -f2,5 | sort -g -k2 | uniq -c | sed 's/^ *//' | awk -v OFS='\t' '{ print $2,$3,$1}' &gt; BC${j}.hits.txt; done
</code></pre>
<p>Now have a look at <code>BC1.hits.txt</code>. It should be a tab-delimited text file with three columns:</p>
<ol>
<li>the reference sequence ID (essential for multi-contig genomes)</li>
<li>the position along that reference</li>
<li>the total number of transposon insertions or "hits" at that position</li>
</ol>
<h3 id="collect-hit-statistics-by-gene">Collect hit statistics by gene</h3>
<p>I am not a Python expert, but in this repo there are two scripts we'll use for converting our <code>.hits.txt</code> data by locus tag, using a Genbank file of the reference genome containing the annotation and coordinates of each gene feature.</p>
<pre><code>mkdir tabulate &amp;&amp; cd tabulate
tabulate_insertions.py ~/tnseq/reference/Geobacter_sulfurreducens_MN1.gbk ../BC1.hits.txt BC1 0 0.05
</code></pre>
<p>The last two numbers specify a percentage to trim off the N- and C-terminus, respectively, of the coding sequence. In past analyses we only consider hits within the first 95% of the total gene length</p>
<p>As the script runs, you'll see a scrolling output of genes that do not contain any TA site hits. What is significant about these genes?</p>
<h3 id="calculate-log2-ratios-to-identify-genes-of-interest">Calculate log~2~ ratios to identify genes of interest</h3>
<p>Two particular loci should jump out at us: GSU0274 and GSU3259. These genes encode inner membrane cytochromes that function at low and high potential, respectively. So, in the BC2 library (low potential electrode outgrowth), we should see very few hits in GSU0274. Likewise, in the BC3 library, we should see very few hits in GSU3259. Let's use the second script for this:</p>
<pre><code>insertion_statistics_by_locustag.py /tnseq/reference/Geobacter_sulfurreducens_MN1.gbk ../BC2.hits.txt GSU0274 0 0.05
insertion_statistics_by_locustag.py /tnseq/reference/Geobacter_sulfurreducens_MN1.gbk ../BC3.hits.txt GSU3259 0 0.05
</code></pre></div>
        </div>

	<footer class="col-md-12">
		<hr>
		
		<p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
	</footer>


        <script src="./js/jquery-1.10.2.min.js"></script>
        <script src="./js/bootstrap-3.0.3.min.js"></script>
        <script src="./js/highlight.pack.js"></script>
        <script src="./js/base.js"></script>
    </body>
</html>